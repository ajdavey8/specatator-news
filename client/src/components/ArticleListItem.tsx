import React, { useContext } from 'react';
import { faStar } from '@fortawesome/free-regular-svg-icons';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styled, { ThemeContext } from 'styled-components';
import { Article } from '../types';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  padding-bottom: ${({ theme }) => theme.spacingM};
  margin-bottom: ${({ theme }) => theme.spacingL};
  border-bottom: 0.5px solid ${({ theme }) => theme.lightGrey};
`

const Author = styled.a`
  color:  ${({ theme }) => theme.primaryRed};
  margin-bottom: ${({ theme }) => theme.spacingS};
  font-style: italic;
  font-size: 18px;
`

const Title = styled.h3`
  font-size: 24px;
`

const Button = styled.button<{
  bookmarked?: boolean;
}>`
  margin: ${({ theme }) => theme.spacingM} 0;
  width: 112px;
  padding: 4px 0;
  font-size: 12px;
  font-family: ${({ theme }) => theme.buttonFont};
  border-radius: 10px;
  border: solid 0.5px ${props => props.bookmarked ? 'rgba(38, 34, 34, 0.15)' : props.theme.primaryRed};
  background-color: ${props => props.bookmarked ? 'rgba(38, 34, 34, 0.05)' : props.theme.secondaryRed};
  color:  ${props => props.bookmarked ? 'rgba(38, 34, 34, 0.4)' : props.theme.primaryRed};
  
  ${(props) => !props.bookmarked && `
    box-shadow: 2px 2px 0 0 ${props.theme.secondaryRed}, inset 2px 2px 0 0 #ffffff;
`}
`

const Image = styled.img`
  width: 100%;

  ${({ theme }) => theme.mediaMedium} {
    height: 160px;
  }

  ${({ theme }) => theme.mediaLarge} {
    height: 160px;
  }
`

interface ArticleListItem extends Article {
  addBookmark: (arg1: number) => void
}

const ArticleListItem = ({ id, title, author, image, bookmarked, addBookmark }: ArticleListItem) => {
  const { green } = useContext(ThemeContext);

  const handleClick = () => {
    addBookmark(id)
  }

  return (
    <Container>
      <Author>{author}</Author>
      <Title>{title}</Title>
      <Button bookmarked={bookmarked} disabled={bookmarked} onClick={handleClick}>
        {bookmarked ?
          (<>ADDED <FontAwesomeIcon icon={faCheck} color={green} /> </>) :
          (<>ADD <FontAwesomeIcon icon={faStar} /> </>)
        }
      </Button>
      <Image src={image} />
    </Container>
  )
}

export default ArticleListItem;
