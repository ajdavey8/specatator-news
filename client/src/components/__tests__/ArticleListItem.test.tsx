import React from 'react';
import { render, fireEvent } from '@testing-library/react'
import ArticleListItem from '../ArticleListItem';
import { ThemeProvider } from 'styled-components';

const addBookmark = jest.fn()

const theme = {
  green: "#29724c",
}

const setup = () => {
  const utils = render(
    <ThemeProvider theme={theme}>
      <ArticleListItem
        id={1}
        author='Alex'
        title='title'
        image='imageURL'
        addBookmark={addBookmark}
      />
    </ThemeProvider>
  )

  return {
    ...utils
  }
}

describe('<ArticleListItem>', () => {
  it('renders with the passed props', () => {
    const { asFragment } = setup()
    expect(asFragment).toMatchSnapshot();
  })

  it('calls the onClick function to add bookmark', () => {
    const { getByText } = setup()

    fireEvent.click(getByText('ADD'))

    expect(addBookmark).toHaveBeenCalledWith(1)
  })

})