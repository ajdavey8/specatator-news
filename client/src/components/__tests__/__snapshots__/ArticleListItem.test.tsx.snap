// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleListItem> renders with the passed props 1`] = `[Function]`;
